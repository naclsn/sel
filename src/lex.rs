//! lexing

use std::iter::{self, Enumerate, Peekable};

use crate::scope::{Location, SourceRef};

#[derive(PartialEq, Debug, Clone)]
pub enum TokenKind {
    Unknown(String),
    Number(f64),
    Bytes(Box<[u8]>),
    Word(String),
    Comma,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
    Equal,
    Def,
    Let,
    Use,
    //LineComment(String),
    //DashComment(String),
    End,
}

#[derive(PartialEq, Debug, Clone)]
pub struct Token(pub Location, pub TokenKind);

/// note: this is an infinite iterator (`next()` is never `None`)
pub struct Lexer<I: Iterator<Item = u8>> {
    stream: Peekable<Enumerate<I>>,
    source: SourceRef,
    last_at: usize,
}

impl<I: Iterator<Item = u8>> Lexer<I> {
    pub(crate) fn new(source: SourceRef, iter: I) -> Self {
        Self {
            stream: iter.enumerate().peekable(),
            source,
            last_at: 0,
        }
    }
}

impl<I: Iterator<Item = u8>> Iterator for Lexer<I> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        use TokenKind::*;

        let Some((at, byte)) = self.stream.find(|c| !c.1.is_ascii_whitespace()) else {
            let range = self.last_at..self.last_at;
            return Some(Token(Location(self.source, range), End));
        };

        let (len, tok) = match byte {
            b':' => {
                let mut doubles = 2;
                let b: Box<[u8]> =
                    iter::from_fn(|| match (self.stream.next(), self.stream.peek()) {
                        (Some((_, b':')), Some((_, b':'))) => {
                            doubles += 1;
                            self.stream.next();
                            Some(b':')
                        }
                        (Some((_, b':')), _) => None,
                        (pair, _) => pair.map(|c| c.1),
                    })
                    .collect();
                (b.len() + doubles, Bytes(b))
            }

            b',' => (1, Comma),
            b'[' => (1, OpenBracket),
            b']' => (1, CloseBracket),
            b'{' => (1, OpenBrace),
            b'}' => (1, CloseBrace),
            b'=' => (1, Equal),

            b'#' => {
                if let Some((_, b'-')) = self.stream.peek() {
                    self.stream.next();
                    let mut m = Vec::new();
                    while {
                        match self.next().expect("unreachable: infinite iterator").1 {
                            OpenBracket => m.push(b'b'),
                            CloseBracket if m.last().is_some_and(|b| b'b' == *b) => _ = m.pop(),
                            OpenBrace => m.push(b'B'),
                            CloseBrace if m.last().is_some_and(|b| b'B' == *b) => _ = m.pop(),
                            End => m.clear(),
                            _ => {}
                        }
                        !m.is_empty()
                    } {}
                    let n = self.next();
                    let Some(Token(_, Comma)) = n else {
                        return n;
                    };
                } else {
                    self.stream.find(|c| {
                        self.last_at = c.0;
                        b'\n' == c.1
                    });
                }
                return self.next();
            }
            b'_' => (1, Word("_".into())),

            c if c.is_ascii_lowercase() || b'-' == c => {
                let b: Vec<u8> = iter::once(c)
                    .chain(iter::from_fn(|| {
                        self.stream
                            .next_if(|c| c.1.is_ascii_lowercase() || b'-' == c.1)
                            .map(|c| c.1)
                    }))
                    .collect();
                (
                    b.len(),
                    match &b[..] {
                        b"def" => Def,
                        b"let" => Let,
                        b"use" => Use,
                        _ => Word(String::from_utf8(b).expect("unreachable: is_ascii_lowercase")),
                    },
                )
            }

            c if c.is_ascii_digit() => {
                let mut r = 0;
                let mut l = 1;

                let (shift, digits) = match (c, self.stream.peek()) {
                    (b'0', Some((_, b'B' | b'b'))) => (1, b"01".as_slice()),
                    (b'0', Some((_, b'O' | b'o'))) => (3, b"01234567".as_slice()),
                    (b'0', Some((_, b'X' | b'x'))) => (4, b"0123456789abcdef".as_slice()),
                    _ => (0, b"0123456789".as_slice()),
                };
                if 0 == shift {
                    r = c as usize - 48;
                } else {
                    l += 1;
                    self.stream.next();
                }

                while let Some(k) = self
                    .stream
                    .peek()
                    .and_then(|c| digits.iter().position(|&k| k == c.1 | 32))
                {
                    l += 1;
                    self.stream.next();
                    r = if 0 == shift {
                        r * 10 + k
                    } else {
                        (r << shift) | k
                    };
                }

                let r = if 0 == shift && self.stream.peek().is_some_and(|c| b'.' == c.1) {
                    l += 2;
                    self.stream.next();
                    let mut d = match self.stream.next() {
                        Some(c) if c.1.is_ascii_digit() => (c.1 - b'0') as usize,
                        _ => {
                            self.last_at = at + l;
                            let t = r.to_string() + ".";
                            return Some(Token(Location(self.source, at..at + l), Unknown(t)));
                        }
                    };
                    let mut w = 10;
                    while let Some(c) = self.stream.peek().map(|c| c.1).filter(u8::is_ascii_digit) {
                        l += 1;
                        self.stream.next();
                        d = d * 10 + (c - b'0') as usize;
                        w *= 10;
                    }
                    r as f64 + (d as f64 / w as f64)
                } else {
                    r as f64
                };

                (l, Number(r))
            }

            c => {
                let chclass = c.is_ascii_alphanumeric();
                let b: Vec<u8> = iter::once(c)
                    .chain(iter::from_fn(|| {
                        self.stream
                            .next_if(|c| {
                                c.1.is_ascii_alphanumeric() == chclass && !c.1.is_ascii_whitespace()
                            })
                            .map(|c| c.1)
                    }))
                    .collect();
                (b.len(), Unknown(String::from_utf8_lossy(&b).into_owned()))
            }
        };

        self.last_at = at + len;
        Some(Token(Location(self.source, at..at + len), tok))
    }
}

#[test]
fn test() {
    use insta::assert_snapshot;

    fn t(script: &[u8]) -> String {
        let mut end = true;
        Lexer::new(0, script.iter().copied())
            .take_while(|t| {
                if !end {
                    false
                } else {
                    end = TokenKind::End != t.1;
                    true
                }
            })
            .fold(String::new(), |mut acc, cur| {
                acc += &format!("{cur:?}\n");
                acc
            })
    }

    assert_snapshot!(t(b""));
    assert_snapshot!(t(b"coucou"));
    assert_snapshot!(t(b"[a 1 b, {w, x, y, z}, 0.5]"));
    assert_snapshot!(t(b":hay: :hey:: not hay: :: :::: fin"));
    assert_snapshot!(t(b"42 :*: 0x2a 0b101010 0o52"));
    assert_snapshot!(t(b"oh-no; don't let use be the def of cat!"));
    assert_snapshot!(t(b"
normal
# comment
normal again
#-comment not comment
#- comment again not comment
a #- [ba ba ba] b
a #- 239084 b
a #- :yeeesldkfj: b
a #- {one, [two], [{three}]} b
a #- {one, #-[two], [{three}]} b
{heyo, #-baba, owieur} # the ',' is also commented
{heyo, baba, #-owieur} # the '}' isn't
"));
}
