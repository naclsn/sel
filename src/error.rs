use std::fmt::{Display, Formatter, Result as FmtResult};

use crate::parse::{Applicable, TokenKind};
use crate::scope::{Location, SourceRegistry};
use crate::types::FrozenType;

// error types {{{
#[derive(PartialEq, Debug)]
pub enum ErrorContext {
    Unmatched {
        open_token: TokenKind,
    },
    CompleteType {
        complete_type: FrozenType,
    },
    TypeListInferredItemType {
        list_item_type: FrozenType,
    },
    AsNthArgToNowTyped {
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromAsNthArgToNowTyped {
        comma_loc: Location,
        nth_arg: usize,
        func: Applicable,
        type_with_curr_args: FrozenType,
    },
    ChainedFromToNotFunc {
        comma_loc: Location,
    },
    AutoCoercedVia {
        func_name: String,
        func_type: FrozenType,
    },
    DeclaredHereWithType {
        with_type: FrozenType,
    },
    LetFallbackTypeMismatch {
        result_type: FrozenType,
        fallback_type: FrozenType,
    },
}

#[derive(PartialEq, Debug)]
pub enum ErrorKind {
    ContextCaused {
        error: Box<Error>,
        because: ErrorContext,
    },
    Unexpected {
        token: TokenKind,
        expected: &'static str,
    },
    UnexpectedDefInScript,
    UnknownName {
        name: String,
        expected_type: FrozenType,
    },
    NotFunc {
        actual_type: FrozenType,
    },
    TooManyArgs {
        nth_arg: usize,
        func: Applicable,
    },
    ExpectedButGot {
        expected: FrozenType,
        actual: FrozenType,
    },
    InfWhereFinExpected,
    NameAlreadyDeclared {
        name: String,
    },
    CouldNotReadFile {
        error: String,
    },
}

#[derive(PartialEq, Debug)]
pub struct Error(pub Location, pub ErrorKind);

#[derive(Default, PartialEq, Debug)]
pub struct ErrorList(pub(crate) Vec<Error>);

impl IntoIterator for ErrorList {
    type Item = Error;
    type IntoIter = <Vec<Error> as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl ErrorList {
    pub(crate) fn push(&mut self, err: Error) {
        self.0.push(err);
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}
// }}}

pub struct Report<'a> {
    registry: &'a SourceRegistry,
    title: String,
    messages: Vec<(Location, String)>,
}

// generate report {{{
impl Error {
    fn ctx_messages(loc: Location, because: &ErrorContext, report: &mut Report) {
        use ErrorContext::*;
        use TokenKind::*;

        let msgs: &[_] = match because {
            Unmatched { open_token } => &[(
                loc,
                format!(
                    "open {} here",
                    match open_token {
                        OpenBracket => '[',
                        OpenBrace => '{',
                        Unknown(_) | Word(_) | Bytes(_) | Number(_) | Comma | CloseBracket
                        | CloseBrace | Equal | Def | Let | Use | Semicolon | End => unreachable!(),
                    }
                ),
            )],
            CompleteType { complete_type } => &[(loc, format!("complete type: {complete_type}"))],
            TypeListInferredItemType { list_item_type } => &[(
                loc,
                format!("list type was inferred to be [{list_item_type}]"),
            )],
            AsNthArgToNowTyped {
                nth_arg,
                func,
                type_with_curr_args,
            } => &[(
                loc,
                format!(
                    "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                    match nth_arg {
                        1 => "st",
                        2 => "nd",
                        3 => "rd",
                        _ => "th",
                    },
                    match func {
                        Applicable::Name(name) => name,
                        Applicable::Bind(_, _, _) => "let binding",
                    },
                ),
            )],
            ChainedFromAsNthArgToNowTyped {
                comma_loc,
                nth_arg,
                func,
                type_with_curr_args,
            } => &[
                (
                    loc,
                    format!(
                        "see parameter in {type_with_curr_args} (overall {nth_arg}{} argument to {})",
                        match nth_arg {
                            1 => "st",
                            2 => "nd",
                            3 => "rd",
                            _ => "th",
                        },
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                    ),
                ),
                (comma_loc.clone(), "chained through here".into()),
            ],
            ChainedFromToNotFunc { comma_loc } => &[
                (loc, "Not a function".into()),
                (comma_loc.clone(), "chained through here".into()),
            ],
            AutoCoercedVia {
                func_name,
                func_type,
            } => &[(
                loc,
                format!("coerced via '{func_name}' (which has type {func_type})"),
            )],
            DeclaredHereWithType { with_type } => &[(loc, format!("here with type {with_type}"))],
            LetFallbackTypeMismatch {
                result_type,
                fallback_type,
            } => &[(
                loc,
                format!("fallback of type {fallback_type} doesn't match result type {result_type}"),
            )],
        };
        report.messages.extend_from_slice(msgs);
    }

    pub fn report<'a>(&self, registry: &'a SourceRegistry) -> Report<'a> {
        use ErrorKind::*;
        use TokenKind::*;

        let loc = self.0.clone();

        match &self.1 {
            ContextCaused { error, because } => {
                let mut r = error.report(registry);
                Error::ctx_messages(loc, because, &mut r);
                r
            }
            Unexpected { token, expected } => Report {
                registry,
                title: "Unexpected token".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Unexpected '{}', expected {expected}",
                        match token {
                            Unknown(t) => t,
                            Word(v) => v,
                            Bytes(_) => "(bytes)",
                            Number(_) => "(number)",
                            Comma => ",",
                            OpenBracket => "[",
                            CloseBracket => "]",
                            OpenBrace => "{",
                            CloseBrace => "}",
                            Equal => "=",
                            Def => "def",
                            Let => "let",
                            Use => "use",
                            Semicolon => ";",
                            End => "end of script",
                        }
                    ),
                )],
            },
            UnexpectedDefInScript => Report {
                registry,
                title: "Unexpected definition".into(),
                messages: vec![(
                    loc,
                    "these should only appear before the script, at the beginning of the file"
                        .into(),
                )],
            },
            UnknownName {
                name,
                expected_type,
            } => Report {
                registry,
                title: "Unknown name".into(),
                messages: vec![(
                    loc,
                    format!("Unknown name '{name}', should be of type {expected_type}"),
                )],
            },
            NotFunc { actual_type } => Report {
                registry,
                title: "Not a function".into(),
                messages: vec![(
                    loc,
                    format!("Expected a function type, but got {actual_type}"),
                )],
            },
            TooManyArgs { nth_arg, func } => Report {
                registry,
                title: "Too many argument".into(),
                messages: vec![(
                    loc,
                    format!(
                        "Too many argument to {}, expected only {}",
                        match func {
                            Applicable::Name(name) => name,
                            Applicable::Bind(_, _, _) => "let binding",
                        },
                        nth_arg - 1
                    ),
                )],
            },
            ExpectedButGot { expected, actual } => Report {
                registry,
                title: "Type mismatch".into(),
                messages: vec![(loc, format!("Expected type {expected}, but got {actual}"))],
            },
            InfWhereFinExpected => Report {
                registry,
                title: "Wrong boundedness".into(),
                messages: vec![(loc, "Expected finite type, but got infinite type".into())],
            },
            NameAlreadyDeclared { name } => Report {
                registry,
                title: "Name used twice".into(),
                messages: vec![(loc, format!("Name {name} was already declared"))],
            },
            CouldNotReadFile { error } => Report {
                registry,
                title: "Could not read file".into(),
                messages: vec![(loc, format!("<< {error} >>"))],
            },
        }
    }
}
// }}}

// display report {{{
impl Display for Report<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        let (Location(file, ref range), _) = self.messages[0];
        let path = self.registry.get_path(file);
        let bytes = self.registry.get_bytes(file);
        let line = 1 + bytes[range.start..].iter().filter(|c| b'\n' == **c).count();
        writeln!(f, "{}:{line}: {}", path.display(), self.title)?;

        for (_, msg) in &self.messages {
            writeln!(f, " -> {}", msg)?;
        }
        Ok(())
    }
}
// }}}
