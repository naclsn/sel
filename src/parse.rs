/// parsing into a CST
use std::iter::Peekable;

use crate::error::{self, Error, ErrorKind};
use crate::lex::{Lexer, Token, TokenKind};
use crate::scope::SourceRef;

macro_rules! TermToken {
    () => {
        TokenKind::Comma
            | TokenKind::CloseBracket
            | TokenKind::CloseBrace
            | TokenKind::Equal // in practice it is always consumed by `parse_value`
            | TokenKind::End
    };
}

/// `top ::= {'use' <bytes> <word> ','} {'def' <word> <bytes> <value> ','} [<script>]`
#[derive(Debug, Clone)]
pub struct Top {
    pub uses: Vec<Use>,
    pub defs: Vec<Def>,
    pub script: Option<Script>,
}
#[derive(Debug, Clone)]
pub struct Use {
    pub path: Vec<u8>,
    pub name: String,
}
#[derive(Debug, Clone)]
pub struct Def {
    pub name: String,
    pub desc: String,
    pub to: Apply,
}

/// `script ::= <apply> {',' <apply>}`
#[derive(Debug, Clone)]
pub struct Script {
    pub head: Apply,
    pub tail: Vec<Apply>,
}

/// `apply ::= (<binding> | <value>) {<value>}`
#[derive(Debug, Clone)]
pub struct Apply {
    pub base: ApplyBase,
    pub args: Vec<Value>,
}
#[derive(Debug, Clone)]
pub enum ApplyBase {
    /// `binding ::= 'let' (<irrefut> <value> | <pattern> <value> <value>)`
    Binding(Pattern, Box<Value>, Box<Value>), // garbage if `pat.is_irrefutable()`
    /// `value ::= <atom> | <subscr> | <list> | <pair>`
    Value(Box<Value>),
}

/// `value ::= <atom> | <subscr> | <list> | <pair>`
#[derive(Debug, Clone)]
pub enum Value {
    /// `subscr ::= '[' <script> ']'`
    Subscr(Script),

    /// `list ::= '{' [<apply> {',' <apply>} [',' [',' <apply>]]] '}'`
    List(Vec<Apply>, Option<Box<Apply>>),
    /// `pair ::= (<atom> | <subscr> | <list>) '=' <value>`
    Pair(Box<Value>, Box<Value>),

    /// `atom ::= <word> | <bytes> | <number>`
    /// `word ::= /[-a-z]+/ | '_'`
    Word(String),
    /// `atom ::= <word> | <bytes> | <number>`
    /// `bytes ::= /:([^:]|::)*:/`
    Bytes(Vec<u8>),
    /// `atom ::= <word> | <bytes> | <number>`
    /// `number ::= /0b[01]+/ | /0o[0-7]+/ | /0x[0-9A-Fa-f]+/ | /[0-9]+(\.[0-9]+)?/`
    Number(f64),
}

/// `irrefut ::= <word> | <irrefut> '=' <irrefut>`
/// `pattern ::= <atom> | <patlist> | <patpair>`
#[derive(Debug, Clone)]
pub enum Pattern {
    /// `atom ::= <word> | <bytes> | <number>`
    /// `word ::= /[-a-z]+/ | '_'`
    Word(String),
    /// `atom ::= <word> | <bytes> | <number>`
    /// `bytes ::= /:([^:]|::)*:/`
    Bytes(Vec<u8>),
    /// `atom ::= <word> | <bytes> | <number>`
    /// `number ::= /0b[01]+/ | /0o[0-7]+/ | /0x[0-9A-Fa-f]+/ | /[0-9]+(\.[0-9]+)?/`
    Number(f64),

    /// `patlist ::= '{' [<pattern> {',' <pattern>} [',' [',' <word>]]] '}'`
    List(Vec<Pattern>, Option<String>),
    /// `patpair ::= (<atom> | <patlist>) '=' <pattern>`
    Pair(Box<Pattern>, Box<Pattern>),
}

impl Pattern {
    pub fn is_irrefutable(&self) -> bool {
        match self {
            Pattern::Word(_) => true,
            Pattern::Pair(fst, snd) => fst.is_irrefutable() && snd.is_irrefutable(),
            Pattern::Number(_) | Pattern::Bytes(_) | Pattern::List(_, _) => false,
        }
    }
}

pub struct Parser<I: Iterator<Item = u8>> {
    lex: Peekable<Lexer<I>>,
    //comments: Vec<..>,
    errors: Vec<Error>,
}

impl<I: Iterator<Item = u8>> Parser<I> {
    pub fn new(source: SourceRef, bytes: I) -> Parser<I> {
        Parser {
            lex: Lexer::new(source, bytes).peekable(),
            errors: Vec::new(),
        }
    }

    pub fn errors(&self) -> &[Error] {
        &self.errors
    }

    fn peek_tok(&mut self) -> &Token {
        self.lex.peek().expect("unreachable: infinite iterator")
    }
    fn next_tok(&mut self) -> Token {
        self.lex.next().expect("unreachable: infinite iterator")
    }
    fn skip_tok(&mut self) {
        self.lex.next();
    }

    pub fn parse_top(&mut self) -> Top {
        Top {
            uses: self.parse_uses(),
            defs: self.parse_defs(),
            script: if let Token(_, TokenKind::End) = self.peek_tok() {
                None
            } else {
                Some(self.parse_script())
            },
        }
    }

    pub fn parse_uses(&mut self) -> Vec<Use> {
        let mut r = Vec::new();
        while let Token(_, TokenKind::Use) = self.peek_tok() {
            self.skip_tok();

            let (_loc, path, name) = match (self.next_tok(), self.next_tok()) {
                (Token(loc, TokenKind::Bytes(file)), Token(_, TokenKind::Word(mut name))) => {
                    if "_" == name {
                        name.clear();
                    } else {
                        name.push('-');
                    }
                    (loc, file, name)
                }
                (Token(_, TokenKind::Bytes(_)), other) | (other, _) => {
                    self.errors.push(error::unexpected(
                        &other,
                        "file path then identifier after 'use' keyword",
                        None,
                    ));
                    continue;
                }
            };

            r.push(Use { path, name });

            match self.peek_tok() {
                Token(_, TokenKind::Comma | TokenKind::End) => self.skip_tok(),
                other @ Token(_, kind) => {
                    let err = error::unexpected(other, "a ',' (because of prior 'use')", None);
                    if matches!(kind, TermToken!()) {
                        self.skip_tok();
                    }
                    self.errors.push(err);
                }
            }
        }
        r
    }

    pub fn parse_defs(&mut self) -> Vec<Def> {
        let mut r = Vec::new();
        while let Token(_, TokenKind::Def) = self.peek_tok() {
            self.skip_tok();

            let (_loc, name, desc) = match (self.next_tok(), self.next_tok()) {
                (Token(loc, TokenKind::Word(name)), Token(_, TokenKind::Bytes(desc))) => {
                    (loc, name, desc)
                }
                (Token(_, TokenKind::Word(_)), other) | (other, _) => {
                    self.errors.push(error::unexpected(
                        &other,
                        "name then description after 'def' keyword",
                        None,
                    ));
                    continue;
                }
            };

            // TODO(wrong todo location): maybe use description to provide more meaningfull var type names

            // TODO: https://peps.python.org/pep-0257/#handling-docstring-indentation

            r.push(Def {
                name,
                desc: String::from_utf8_lossy(&desc).trim().into(),
                to: self.parse_apply(),
            });

            match self.peek_tok() {
                Token(_, TokenKind::Comma | TokenKind::End) => self.skip_tok(),
                other @ Token(_, kind) => {
                    let err = error::unexpected(other, "a ',' (because of prior 'def')", None);
                    if matches!(kind, TermToken!()) {
                        self.skip_tok();
                    }
                    self.errors.push(err);
                }
            }
        }
        r
    }

    pub fn parse_script(&mut self) -> Script {
        let head = self.parse_apply();
        let mut tail = Vec::new();
        while let Token(_comma_loc, TokenKind::Comma) = self.peek_tok() {
            self.skip_tok();
            tail.push(self.parse_apply());
        }
        Script { head, tail }
    }

    pub fn parse_apply(&mut self) -> Apply {
        let base = match self.peek_tok() {
            Token(let_loc, TokenKind::Let) => {
                let let_loc = let_loc.clone();
                self.skip_tok();

                let pat = self.parse_pattern();
                let res = Box::new(self.parse_value());

                // if irrefutable, no fallback
                let alt = if pat.is_irrefutable() {
                    Box::new(Value::Number(0.0))
                } else {
                    let plen = self.errors.len();
                    let value = self.parse_value();

                    // try to catch and report a common mistake
                    if plen + 1 == self.errors.len() {
                        if let err @ Error(_, ErrorKind::Unexpected { .. }) =
                            self.errors.pop().expect("len >0")
                        {
                            let err = error::context_fallback_required(let_loc, err);
                            self.errors.push(err);
                        }
                    }

                    Box::new(value)
                };

                ApplyBase::Binding(pat, res, alt)
            }

            _ => ApplyBase::Value(Box::new(self.parse_value())),
        };

        let mut args = Vec::new();
        while !matches!(self.peek_tok(), Token(_, TermToken!())) {
            args.push(self.parse_value());
        }

        Apply { base, args }
    }

    fn parse_subscr_content(&mut self, first_token: &Token) -> Script {
        let subscr = self.parse_script();

        match self.peek_tok() {
            Token(_close_loc, TokenKind::CloseBracket) => {
                //loc.1.end = close_loc.1.end;
                self.skip_tok()
            }
            other @ Token(_, kind) => {
                let skip = TokenKind::Comma == *kind;
                let err =
                    error::unexpected(other, "next argument or closing ']'", Some(&first_token));
                self.errors.push(err);
                if skip {
                    self.skip_tok();
                }
            }
        }

        subscr
    }

    fn parse_list_content(&mut self, first_token: &Token) -> (Vec<Apply>, Option<Box<Apply>>) {
        let mut items = Vec::new();
        let mut rest = None;

        while match self.peek_tok() {
            Token(_close_loc, TokenKind::CloseBrace) => {
                //loc.1.end = close_loc.1.end;
                self.skip_tok();
                false
            }

            #[allow(unreachable_patterns, reason = "because of 'CloseBrace'")]
            term @ Token(_, TermToken!()) => {
                let err = error::unexpected(term, "next item or closing '}'", Some(first_token));
                self.errors.push(err);
                false
            }

            _ => true,
        } {
            items.push(self.parse_apply());
            // parse_apply ends on a TermToken

            if matches!(self.peek_tok(), Token(_, TokenKind::Comma)) {
                self.skip_tok();
                // { ... ,,
                if matches!(self.peek_tok(), Token(_, TokenKind::Comma)) {
                    self.skip_tok();
                    rest = Some(Box::new(self.parse_apply()));
                    // parse_apply ends on a TermToken

                    match self.peek_tok() {
                        Token(_close_loc, TokenKind::CloseBrace) => {
                            //loc.1.end = close_loc.1.end;
                            self.skip_tok();
                        }
                        other => {
                            let err = error::unexpected(
                                other,
                                "closing '}' after ',,'",
                                Some(&first_token),
                            );
                            self.errors.push(err);
                        }
                    }

                    break;
                } // if ',,'
            } // if ','
        } // while

        (items, rest)
    }

    pub fn parse_value(&mut self) -> Value {
        // don't consume a TermToken at this point
        if let term @ Token(_loc, TermToken!()) = self.peek_tok() {
            let err = error::unexpected(term, "a value", None);
            self.errors.push(err);
            return Value::Word("?".into());
        }

        let first_token = self.next_tok();

        //let mut loc = first_token.0.clone();
        let value = match first_token.1 {
            TokenKind::OpenBracket => {
                let subscr = self.parse_subscr_content(&first_token);
                Value::Subscr(subscr)
            }

            TokenKind::OpenBrace => {
                let (items, rest) = self.parse_list_content(&first_token);
                Value::List(items, rest)
            }

            TokenKind::Word(w) => Value::Word(w),
            TokenKind::Bytes(b) => Value::Bytes(b),
            TokenKind::Number(n) => Value::Number(n),

            TokenKind::Def | TokenKind::Let | TokenKind::Use | TokenKind::Unknown(_) => {
                self.errors
                    .push(error::unexpected(&first_token, "a value", None));
                Value::Word(
                    match &first_token.1 {
                        TokenKind::Def => "?def",
                        TokenKind::Let => "?let",
                        TokenKind::Use => "?use",
                        TokenKind::Unknown(t) => t,
                        _ => unreachable!(),
                    }
                    .into(),
                )
            }

            TermToken!() => unreachable!(),
        };

        if let Token(_, TokenKind::Equal) = self.peek_tok() {
            self.skip_tok();
            Value::Pair(Box::new(value), Box::new(self.parse_value()))
        } else {
            value
        }
    }

    pub fn parse_pattern(&mut self) -> Pattern {
        // don't consume a TermToken at this point
        if let term @ Token(_loc, TokenKind::OpenBracket | TermToken!()) = self.peek_tok() {
            let err = error::unexpected(term, "a pattern", None);
            self.errors.push(err);
            return Pattern::Word("?".into());
        }

        let first_token = self.next_tok();

        let pattern = match first_token.1 {
            TokenKind::OpenBrace => {
                let mut items = Vec::new();
                let mut rest = None;

                loop {
                    match self.peek_tok() {
                        Token(_close_loc, TokenKind::CloseBrace) => {
                            self.skip_tok();
                            break;
                        }

                        #[allow(unreachable_patterns, reason = "because of 'CloseBrace'")]
                        term @ Token(_, TermToken!()) => {
                            let err = error::unexpected(
                                term,
                                "next item or closing '}'",
                                Some(&first_token),
                            );
                            self.errors.push(err);
                            break;
                        }

                        _ => (), // continue
                    }

                    items.push(self.parse_pattern());

                    match self.peek_tok() {
                        Token(_, TokenKind::Comma | TokenKind::CloseBrace) => (),
                        other => {
                            let err = error::unexpected(
                                other,
                                "',' between items or closing '}'",
                                Some(&first_token),
                            );
                            self.errors.push(err);
                        }
                    }

                    // { ... ,
                    if matches!(self.peek_tok(), Token(_, TokenKind::Comma)) {
                        self.skip_tok();
                        if !matches!(self.peek_tok(), Token(_, TokenKind::Comma)) {
                            continue;
                        }
                        self.skip_tok();
                        // { ... ,,

                        rest = Some(match self.next_tok() {
                            Token(_, TokenKind::Word(w)) => w,
                            other => {
                                self.errors.push(error::unexpected(
                                    &other,
                                    "a word then closing '}' after ',,'",
                                    Some(&first_token),
                                ));
                                match &other.1 {
                                    TokenKind::CloseBrace => break,
                                    TokenKind::Def => "?def",
                                    TokenKind::Let => "?let",
                                    TokenKind::Use => "?use",
                                    TokenKind::Unknown(t) => t,
                                    _ => "?",
                                }
                                .into()
                            }
                        });

                        match self.peek_tok() {
                            Token(_close_loc, TokenKind::CloseBrace) => {
                                self.skip_tok();
                            }
                            other => {
                                let err = error::unexpected(
                                    other,
                                    "closing '}' after ',,'",
                                    Some(&first_token),
                                );
                                self.errors.push(err);
                                // should this branch continue or break? it can be 1 of 2 situations:
                                // - the intended '}' /is present/ later on, so continue is better
                                // - there is no corresponding '}' *at all*, in which case the risk is to eat the whole file...
                            }
                        }
                        break;
                    } // if ',', then !!','
                } // loop

                Pattern::List(items, rest)
            }

            TokenKind::Word(w) => Pattern::Word(w),
            TokenKind::Bytes(b) => Pattern::Bytes(b),
            TokenKind::Number(n) => Pattern::Number(n),

            TokenKind::Def | TokenKind::Let | TokenKind::Use | TokenKind::Unknown(_) => {
                self.errors
                    .push(error::unexpected(&first_token, "a pattern", None));
                Pattern::Word(
                    match &first_token.1 {
                        TokenKind::Def => "?def",
                        TokenKind::Let => "?let",
                        TokenKind::Use => "?use",
                        TokenKind::Unknown(t) => t,
                        _ => unreachable!(),
                    }
                    .into(),
                )
            }

            TokenKind::OpenBracket | TermToken!() => unreachable!(),
        };

        if let Token(_, TokenKind::Equal) = self.peek_tok() {
            self.skip_tok();
            Pattern::Pair(Box::new(pattern), Box::new(self.parse_pattern()))
        } else {
            pattern
        }
    }
}

#[test]
fn test() {
    use crate::scope::SourceRegistry;
    use insta::assert_debug_snapshot;

    fn t(script: &[u8]) -> (Top, Box<[Error]>) {
        let mut registry = SourceRegistry::default();
        let source = registry.add_bytes("<test>", script.iter().copied());
        let bytes = &registry.get(source).bytes;
        let mut parser = Parser::new(source, bytes.iter().copied());
        (parser.parse_top(), parser.errors.into_boxed_slice())
    }

    assert_debug_snapshot!(t(b""));
    assert_debug_snapshot!(t(b"-"));
    assert_debug_snapshot!(t(b"- 2"));
    assert_debug_snapshot!(t(b"2 -"));
    assert_debug_snapshot!(t(b"-, split:-:, map[tonum, add1, tostr], join:+:"));
    assert_debug_snapshot!(t(b"tonum, add234121, tostr, ln"));
    assert_debug_snapshot!(t(b"[tonum, add234121, tostr] :13242:"));
    assert_debug_snapshot!(t(b"{repeat 1, {}}"));
    assert_debug_snapshot!(t(b"{repeat 1,, {}}"));
    assert_debug_snapshot!(t(b"{0,, repeat 1}"));
    assert_debug_snapshot!(t(b"a=b"));
    assert_debug_snapshot!(t(b"a=b=c"));
    assert_debug_snapshot!(t(b"a=[b=c]"));
    assert_debug_snapshot!(t(b"[a=b]=c"));
    assert_debug_snapshot!(t(b"let a let b a b a # note: is syntax error"));
    assert_debug_snapshot!(t(b"let {a b, c} 0"));
    // not quite a syntax error but only out of luck:
    // ```sel
    // [let a
    //     [[let b
    //         a
    //     ]                           # b -> a
    //         [panic: unreachable:]]  # a
    // ]                               # a -> a
    //     [panic: unreachable:]       # a
    // ```
    assert_debug_snapshot!(t(b"
let a
    [let b
        a
        [panic: unreachable:]]
    [panic: unreachable:]
"));
}
