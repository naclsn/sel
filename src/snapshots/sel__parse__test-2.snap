---
source: src/parse.rs
expression: "t(b\"-\")"
---
(
    Top {
        uses: [],
        defs: [],
        script: Some(
            Script {
                head: Apply {
                    base: Value(
                        Word(
                            "-",
                        ),
                    ),
                    args: [],
                },
                tail: [],
            },
        ),
    },
    [],
)
